drivers/audio/wm8904.c

```diff
@@ -0,0 +1,728 @@
+		return -EINVAL;
+	}
+
+	wm8904_read_reg(dev, WM8904_REG_CLK_RATES_0, &mclkDiv);
+	fs = (mclk >> (mclkDiv & 0x1U)) / cfg->i2s.frame_clk_freq;
+
+	switch (fs) {
+	case 64:
+		wmfs_ratio = kWM8904_FsRatio64X;
+		break;
+	case 128:
+		wmfs_ratio = kWM8904_FsRatio128X;
+		break;
+	case 192:
+		wmfs_ratio = kWM8904_FsRatio192X;
+		break;
+	case 256:
+		wmfs_ratio = kWM8904_FsRatio256X;
+		break;
+	case 384:
+		wmfs_ratio = kWM8904_FsRatio384X;
+		break;
+	case 512:
+		wmfs_ratio = kWM8904_FsRatio512X;
+		break;
+	case 768:
+		wmfs_ratio = kWM8904_FsRatio768X;
+		break;
+	case 1024:
+		wmfs_ratio = kWM8904_FsRatio1024X;
+		break;
+	case 1408:
+		wmfs_ratio = kWM8904_FsRatio1408X;
+		break;
+	case 1536:
+		wmfs_ratio = kWM8904_FsRatio1536X;
+		break;
+	default:
+		LOG_WRN("Invalid Fs ratio: %d", fs);
+		return -EINVAL;
+	}
+
+	/* Disable SYSCLK */
+	wm8904_write_reg(dev, WM8904_REG_CLK_RATES_2, 0x00);
+
+	/* Set Clock ratio and sample rate */
+	wm8904_write_reg(dev, WM8904_REG_CLK_RATES_1,
+			 ((wmfs_ratio) << 10U) | (uint16_t)(wm_sample_rate));
+
+	switch (cfg->i2s.word_size) {
+	case 16:
+		word_size = 0;
+		break;
+	case 20:
+		word_size = 1;
+		break;
+	case 24:
+		word_size = 2;
+		break;
+	case 32:
+		word_size = 3;
+		break;
+	default:
+		LOG_ERR("Word size %d bits not supported; falling back to 16 bits",
+			cfg->i2s.word_size);
+		word_size = 0;
+		break;
+	}
+	/* Set bit resolution */
+	wm8904_update_reg(dev, WM8904_REG_AUDIO_IF_1, (0x000CU), ((uint16_t)(word_size) << 2U));
+
+	/* Enable SYSCLK */
+	wm8904_write_reg(dev, WM8904_REG_CLK_RATES_2, 0x1007);
+	return 0;
+}
+
+static int wm8904_out_volume_config(const struct device *dev, audio_channel_t channel, int volume)
+{
+	/* Set volume values with VU = 0 */
+	const uint16_t val = WM8904_REGVAL_OUT_VOL(0, 0, 1, volume);
+	const uint16_t mask = WM8904_REGMASK_OUT_VU
+		| WM8904_REGMASK_OUT_ZC
+		| WM8904_REGMASK_OUT_VOL;
+
+	switch (channel) {
+	case AUDIO_CHANNEL_FRONT_LEFT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT2_LEFT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_FRONT_RIGHT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT2_RIGHT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_HEADPHONE_LEFT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_LEFT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_HEADPHONE_RIGHT:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_RIGHT, mask, val);
+		return 0;
+
+	case AUDIO_CHANNEL_ALL:
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_LEFT, mask, val);
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT1_RIGHT, mask, val);
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT2_LEFT, mask, val);
+		wm8904_update_reg(dev, WM8904_REG_ANALOG_OUT2_RIGHT, mask, val);
+		return 0;
+
+	default:
+		return -EINVAL;
+	}
+}
+
+static int wm8904_out_mute_config(const struct device *dev, audio_channel_t channel, bool mute)
+{
+	const uint16_t val = WM8904_REGVAL_OUT_VOL(mute, 0, 0, 0);
+	const uint16_t mask = WM8904_REGMASK_OUT_MUTE;
+
+	switch (channel) {
```

Comment:
This `switch' is the same as in `wm8904_out_mute_config()`. Wouldn't it make sense to use a common part of code for both functions?